src/main/scala/rover/rdo/AtomicObjectState.scala

```diff
@@ -1,22 +1,43 @@
 package rover.rdo
 
-abstract class AtomicObjectState[A <: AtomicObjectState[A]] extends ObjectState {
-	// binding this to a self-type A
-	this: A =>
-
-	private var ops: List[Op] = List()
+import rover.rdo.client.{Log, LogRecord}
 
+// TODO: make ctor private
+class AtomicObjectState[A](private val value: A, private[rdo] val log: Log[A]) extends ObjectState {
 	type Op = A => A
 
-	def applyOp(operation: Op): A = {
+	def immutableState: A = value
+
+	def applyOp(operation: Op): AtomicObjectState[A] = {
 		// Operation must apply itself to the state
 		// but we want the state to take in the operations
 		// so that the framework can record the op
-		val result = operation.apply(this)
+		val result = operation.apply(this.value)
+
+		// Record the operation in the Log
+		val updatedLog = log.appended(LogRecord(value, operation, result))
+
+		return new AtomicObjectState[A](result, updatedLog)
+	}
 
-		// Record the operation in the "log" or "event stream" of operations
-		this.ops = this.ops :+ operation
+	override def equals(that: Any): Boolean = {
+		that match{
+			case that: AtomicObjectState[A] => this.immutableState == that.immutableState
```

Comment:
Does comparing the two with a plain `==` actually work? Does this not just compare the pointers?

Reply:
Yeah it works; i have override the equals method, just to assess equality in terms of the immutable states